### PR TITLE
Fixes #32052 - Dont run actions on disabled services

### PR DESCRIPTION
**Pull request by upadhyeammit**
This change skips to run start, stop, restart and status action on services if those are disabled.

### CODE REVIEW

**Comment by upadhyeammit** [comment]
There can be few different approaches to handle this,

actions = %w[start, stop, restart, status]

* Change pulp2 feature and limit that when pulp_resource_manager service is installed and enabled, very similar to pulpcore feature. If we do it this way then user wont be able to do any actions on pulp2 services from the foreman-maintain unless services are enabled using `systemctl`. This means services wont be listed `service list` command and not available for enabling. I am not in favor of this because on older versions its possible that these services are required however if disabled mistakenly user wont know the correct status of services.
* Go with implementation of this pr. The catch here is, `systemctl` allows to run `actions` on services even when those are disabled? Maybe some people wont like this as its not very similar to standard `systemctl` way? In this change user can still see which services are disabled and can enable it.
* Come up with new flag for service commands, something like --enabled-only. By default this will be false, which means we retain the existing implementation. This flag can be used as required, which means changes required in Katello pipelines and the change will be required until system is in Pulp2 to Pulp3 migration phase? As later on there will be only Pulp3 services. 

@ekohl @jlsherrill what do you think?

**Comment by ehelms** [comment]
You have an accidental inclusion of a built gem binary in the PR.

For what this is used for, and the particular case trying to be solved I think this makes sense. Avoiding any actions on disabled services seems like the right move as they should only be disabled by our software. If a user accidentally disables a service, they will have bigger problems than what this command is reporting and I don't think we should account for users doing something like that.

**Review by ehelms — APPROVED** [review]

**Review by ekohl — COMMENTED** [review]
I'm just going to document how I went through this code.

First of all, I needed to look up the `enabled?` definition. This appears to be:
https://github.com/theforeman/foreman_maintain/blob/4885ae888a8b514c3939254e32c7ad6dcb35bd22/lib/foreman_maintain/utils/service/systemd.rb#L61-L65

Which really calls:
https://github.com/theforeman/foreman_maintain/blob/4885ae888a8b514c3939254e32c7ad6dcb35bd22/lib/foreman_maintain/utils/service/systemd.rb#L73-L75

It's interesting that the `ForemanMaintain::Utils::Service::Abstract` does not define this. Effectively it means that on non-systemd it will fail. Luckily we don't have that anymore, but it's a good thing to be aware of.

If we read the systemctl man page, we get a nice table of statuses. What mostly jumps out to me is `enabled-runtime` as a possible return status when it's enabled via `/run/systemd/system/` instead of `/etc/systemd/system/`. I don't think we use that today, but this can be a potential failure mode.

On a related note, the `exist?` definition doesn't deal with this either and can also fail on `masked` (and `masked-runtime`) where a service does exist but is prevented from starting.

Possible use cases for -runtime would be maintenance mode. Services could be masked in runtime only so that they are started again after a reboot but disallow starting it during the maintenance window. This could be used together with `systemctl isolate` to let systemd manage all services. It is possible that `systemctl rescue` is sufficient. This would also inform systemd it's in maintenance mode. These are just thoughts about embracing system-provided utilities for this and possible alignment. At this point it looks like the code would be compatible with it.

My next question is what it affects. There are many layers in foreman-maintain and it's a bit hard to follow. It would be nice if there were some tests to show some scenarios.

Something else I noticed is that it uses `systemctl status` but according to the same `systemctl` man page that's really intended for humans. Computers should use `systemctl show`, possibly together with `--property`. This also gives you the property `UnitFileState` which appears to be the same as `systemctl is-enabled`. It means you can bulk query all services instead of invoking `systemctl` for every service. This may lead to a lot less process overhead resulting in faster execution.

I also realize that that's probably a larger rewrite. It is roughly the approach I took [in foreman-installer](https://github.com/theforeman/foreman-installer/blob/fdc20de529ebcaba1708bf9afa43a23e96a7858a/hooks/boot/04-services.rb#L30) but there I use `list-units` since I was interested in what's available and what's running.

For now I think this goes in the right direction but I have some implementation details inline that I would like to see addressed. Note that addressed can also mean that you give an explanation why you disagree with me.

File: definitions/features/service.rb [hunk 1]
```diff
@@ -61,7 +61,7 @@ def use_system_service(action, options, spinner)
   def run_action_on_services(action, options, spinner)
     status = 0
     failed_services = []
-    filtered_services(options).each_value do |group|
+    filtered_services(options, action).each_value do |group|
```
[the remark this answers is not in the document]
Comment (upadhyeammit) [reply definitions/features/service.rb:64]: As per the rubocop optional arguments should be the last and so `filtered_services` has `action` as last argument. But then for `filter_services` its not optional hence I kept it by alphanumerically sorted?

**Comment by ehelms** [comment]
Changes LGTM

**Comment by ehelms** [comment]
@ekohl Can you please re-review so we can get pipelines unblocked?

**Comment by ares** [comment]
meanwhile needs rebase :-(

File: definitions/features/service.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (ekohl) [reply definitions/features/service.rb:65]: I would suggest to always add new arguments to methods as the last one. That's the safest because you don't need to re-order every call to it. In this case it also gives you a more consistent API: filtered_services is (options, action) and filter_services would be (service_list, options, action).

Now in this case it's not as critical since it's a required argument, but if it was an optional argument then it's very easy to introduce bugs because no exception would be raised. That's why I'll leave it up to you if you want to make it `def filter_services(options, service_list, action)` or not. However, I do hope you understand why I made the suggestion.

**Review by ekohl — APPROVED** [review]
Per https://github.com/theforeman/foreman_maintain/pull/456/files#r601416907 I'll leave it up to you to what you want to do.

File: definitions/features/service.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (ekohl) [reply definitions/features/service.rb:65]: Oh, I now see my bad, but I meant `def filter_services(service_list, options, action)`. My apologies!
